js/src/edit-block/hooks/useField.js

```diff
@@ -54,58 +104,168 @@ const useField = () => {
+	 * So renaming a field involves changing that key
+	 * and the field's name property.
+	 *
+	 * @param {Object} fields The fields from which to rename a field.
+	 * @param {string} previousName The previous field name (slug).
+	 * @param {string} newName The new field name (slug).
+	 * @return {Object} The fields with the field renamed.
 	 */
-	const getField = useCallback( ( fieldName ) => {
-		return fieldName ? block.fields[ fieldName ] : {};
+	const changeFieldName = ( fields, previousName, newName ) => {
+		fields[ newName ] = { ...fields[ previousName ], name: newName };
+		delete fields[ previousName ];
+		return fields;
+	};
+
+	/**
+	 * Gets the fields for either the editor or inspector.
+	 *
+	 * @param {string} location The location, like 'editor', or 'inspector'.
+	 * @return {Array} The fields with the given location.
+	 */
+	const getFieldsForLocation = useCallback( ( location ) => {
+		if ( ! block || ! block.fields ) {
+			return null;
+		}
+
+		return getFieldsAsArray( block.fields ).filter( ( field ) => {
+			if ( 'editor' === location ) {
+				return ! field.location || 'editor' === field.location;
+			}
+
+			return field.location === 'inspector';
+		} );
 	}, [ block ] );
 
+	/**
+	 * Moves a field to another location, and sets the correct order properties.
+	 *
+	 * @param {number} moveFrom The index of the field to move.
+	 * @param {number} moveTo The index that the field should be moved to.
+	 * @param {string} currentLocation The current field's location, like 'editor'.
+	 */
+	const changeFieldLocation = useCallback( ( fields, fieldName, newLocation ) => {
+		const fieldToMove = fields[ fieldName ];
+		const previousLocation = fieldToMove.location;
+
+		const previousLocationFields = getFieldsForLocation( previousLocation );
+		const fieldsWithoutMovedField = previousLocationFields.filter( ( field ) => {
+			return field.name !== fieldName;
+		} );
+
+		const newLocationFields = getFieldsForLocation( newLocation );
+		newLocationFields.push( fieldToMove );
+
+		return getFieldsAsObject( [
+			...setCorrectOrderForFields( fieldsWithoutMovedField ),
+			...setCorrectOrderForFields( newLocationFields ),
+		] );
+	}, [ getFieldsForLocation ] );
+
 	/**
 	 * Changes a field setting.
 	 *
 	 * @param {string} settingKey The key of the setting, like 'label' or 'placeholder'.
 	 * @param {any} newSettingValue The new setting value.
 	 */
-	const changeFieldSetting = useCallback( ( fieldName, settingKey, newSettingValue ) => {
-		fullBlock[ blockNameWithNamespace ].fields[ fieldName ][ settingKey ] = newSettingValue;
+	const changeFieldSettings = useCallback( ( fieldName, newSettings ) => {
+		if ( newSettings.location ) {
+			fullBlock[ blockNameWithNameSpace ].fields = changeFieldLocation(
+				fullBlock[ blockNameWithNameSpace ].fields,
+				fieldName,
+				newSettings.location
+			);
+		}
+
+		if ( newSettings.hasOwnProperty( 'name' ) ) {
+			fullBlock[ blockNameWithNameSpace ].fields = changeFieldName(
```

Comment:
Changing the name of a field is complex because the name is the key in the `fields object, as well as the name property:

<img width="627" alt="field-name" src="https://user-images.githubusercontent.com/4063887/100783637-a720e900-33d3-11eb-9c65-98973972a84f.png">